src/Microsoft.Extensions.Logging.EventSource/EventSourceLoggerProvider.cs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.Tracing;
+
+namespace Microsoft.Extensions.Logging.EventSourceLogger
+{
+    /// <summary>
+    /// The provider for the <see cref="EventSourceLogger"/>.
+    /// </summary>
+    internal class EventSourceLoggerProvider : ILoggerProvider
+    {
+        // A small integer that uniquely identifies the LoggerFactory assoicated with this LoggingProvider.
+        // Zero is illegal (it means we are uninitialized), and have to be added to the factory. 
+        private int _factoryID;
+
+        private ILoggerFactory _loggerFactory;
+        private LogLevel _defaultLevel;
+        private string _filterSpec;
+        private EventSourceLogger _loggers; // Linked list of loggers that I have created
+
+        public EventSourceLoggerProvider(ILoggerFactory loggerFactory, EventSourceLoggerProvider next = null)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+            _loggerFactory = loggerFactory;
+            Next = next;
+        }
+
+        public readonly EventSourceLoggerProvider Next;
+
+        /// <summary>
+        /// <inheritdoc />
+        /// </summary>
+        public ILogger CreateLogger(string categoryName)
+        {
+            var newLogger = _loggers = new EventSourceLogger(categoryName, _factoryID, _loggers);
```

Comment:
This code is thread safe because the only caller (LoggingEventSource) acquires a lock on the only instance before calling into EventSourceLoggerProvider.

The linked list is there because we need to respond to dynamic configuration change commands sent through EventSource/ETW infrastructure (it all starts at LoggingEventSource.OnEventCommand, which is also thread-safe BTW).

Bottom line, we are bridging ILogger world with EventSource/ETW world. The contract for this bridge should follow ILogger patterns. This is why Vance and I followed your request of making the bridge "off by default". However _the implementation and behavior_ of the bridge follows EventSource/ETW patterns. It is not realistic to expect that this logger implementation will look like any other logger, just pushing data into different place. I think this gives users the best of both worlds, as long as contracts & expectactions from both worlds are satisified